### DISCUSSION TITLE
Script to fuzzy select a channel, update the twt config, stream video and chatter

### DISCUSSION THREAD

**Discussion by shmup** [Xithrius/twitch-tui#186] [Show and tell]
Howdy everyone! I wanted to share a script that works for my process. I save this script as `twitch`.

In a nutshell:

1. Run `twitch` and choose a channel using [fzy](https://github.com/jhawthorn/fzy) as a fuzzy filter.
2. Updates the `$HOME/.config/twt/config.toml` channel with the selection.
3. Streams the `twitch.tv/<selection>` using [streamlink](https://streamlink.github.io/) and [mpv](https://mpv.io/).
4. Runs `twt` to peep the chatter.

[Screenshot](https://user-images.githubusercontent.com/118710/181683886-58698130-daa8-42f1-8c14-bd2bef3b395b.png) and a [shell demo](https://asciinema.org/a/sf5Vr0BBarrY8zgf85aIc3P7b). Pretend you can see the video opening. :innocent:

```bash
#!/usr/bin/env bash

# $HOME/bin/twitch
# fuzzy filtering https://github.com/jhawthorn/fzy
# updates twt config for streaming chat https://github.com/Xithrius/twitch-tui
# streams via mpv and https://github.com/streamlink/streamlink
# runs twt

TWITCHES=(
	"oldtimeycomputershow"
	"dkutelevision"
	"pisshospital"
	"wondervillenyc"
	"sickovision"
	"forgotten_vcr"
	"thatweekinsnl"
	"gamesdonequick"
)

PLAYER="mpv"
PLAYER_ARGS="--really-quiet"
CONFIG=$HOME/.config/twt/config.toml
SELECTION=$(printf "%s\n" "${TWITCHES[@]}" | fzy)

[[ -z $SELECTION ]] && echo "xoxo" && exit 0

# update the twt config w/ the selected channel
sed "s/channel = \".*\"/channel = \"$SELECTION\"/" -i "$CONFIG"

# run --quiet[ly] in the background
streamlink \
	--player "$PLAYER" \
	--player-args "$PLAYER_ARGS" \
	--quiet \
	--default-stream best \
	--retry-streams 3 \
	--url "twitch.tv/$SELECTION" &

# run twt
twt
```

**Comment by shmup** [comment]
I'd like to eventually use the API to get a list of channels I'm following that are _active and live_. Currently I pop into my browser first to see what's streaming.

**Reply by Xithrius** [reply]
This is tracked in #147, but will probably be implemented in 2.1, since I need to get 2.0 out sometime in the near future. The last stable release was in 2021, being 1.6.0.

**Comment by Xithrius** [comment]
The main branch currently implements storage, so items such as previous mentions, and switched-to channels are stored in a JSON file.

This allows the user to hit `s`, and start typing in a channel name. If they've visited it before and the `channels` option in the `[storage]` section is set to true, then they can hit `tab` to complete the channel name.

Same goes for mentions.

**Reply by shmup** [reply]
Very good. Yeah, when I realized I was on `1.6` and you had implemented much more by then, I've switched to `2.0.0-alpha.7`.

Didn't know about `s` yet though. Screenshot for posterity:

![image](https://user-images.githubusercontent.com/118710/181684175-3e845039-eb83-46d0-8997-78b1c8a51ae0.png)

**Reply by Xithrius** [reply]
There are still a bunch of things that are unstable about the alphas, since well, they're alphas.

There's a table for all keybinds that's usually up-to-date in the README. Hitting `?` in the program should yield the same result.

**Reply by Xithrius** [reply]
You should go through all the alpha releases, since they're definitely a metric ton of features that you're missing out on. Scrolling (with the mouse wheel) for some reason took me ages to implement, and it ended up being like 10 lines.

**Reply by Xithrius** [reply]
If you find something that's undocumented, please let me know in an issue!

**Reply by shmup** [reply]
Hm, `?` is not doing that currently, but you've already warned me heh: "since well, they're alphas".

**Reply by Xithrius** [reply]
The keybind should just show a popup of keybinds.

<img width="726" alt="image" src="https://user-images.githubusercontent.com/15021300/181684656-86ae9c7b-6f04-4b59-875e-0d204cf4a40a.png">

**Reply by shmup** [reply]
~~Oh, little did I know that it's just locked up. No more chat streaming. That'd explain it, though. I'll investigate and see what's what. I realized I can't even `q` the process.~~

Is `--log-file` just for chat? Any `--verbose` logging right now?

**Reply by shmup** [reply]
Interesting. I killed one of my tmux panes and see all was well, but the popup was out of view, so I didn't realize _that_ was "locking" the UI, though not really locking. I need to investigate what happened. It could be me, could be something interesting with a small height viewport.

**Reply by Xithrius** [reply]
Currently logging is just for the developers, and no verbose stuff has been added, and for nothing in specific. `info` will be the default log level, and `verbose` will give `debug`.

Now tracking in #187.

**Reply by shmup** [reply]
Figured it out:

![twt](https://user-images.githubusercontent.com/118710/181685201-0b5eed10-36d5-4f86-812c-fb8b681c6345.gif)

**Reply by Xithrius** [reply]
I'm going to be doing a lot of modularity in the near future, so the keybinds window will most likely not be a popup.

This will also allow scrolling in said window.

**Reply by shmup** [reply]
Sounds good. I'm just glad I understand what was going on, heh. `?` is a nice window of info.

**Reply by Xithrius** [reply]
:D

**Comment by Xithrius** [comment]
Hey @shmup, been a while. There have been many changes to this project, mostly including the backend for UI creation/handling, and improvements upon the storage of recent actions by the user (channel switching, mentions, etc). I realize that your script has a hard-coded list of channels, but I think this could be improved upon.

Since there's a dashboard now, it's possible for the user to have a default set of channels they can select from using the [`start_screen_channels`](https://github.com/Xithrius/twitch-tui/blob/main/default-config.toml#L67) config variable. Perhaps the bash script can extract these values, and use them as the previously hard-coded ones.

Along with this, there is a `storage.json` file that is generated at `~/.config/twt/storage.json`, which has all the recent channels and mentions that the user has done (these can be disabled with [some config options](https://github.com/Xithrius/twitch-tui/blob/main/default-config.toml#L21-L25)). By using `jq`, maybe the bash script can use the stored recent channels as well.

Any thoughts on these ideas?

Thanks!

**Comment by Xithrius** [comment]
Managing and interacting with the user's follower list is being done in #419. I'll close this for the time being.
